sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,500 @@
+    if (constructor != null) {
+      return constructor;
+    }
+    Class<? extends DoFnInvoker<?, ?>> invokerClass = generateInvokerClass(signature);
+    try {
+      constructor = invokerClass.getConstructor(fnClass);
+    } catch (IllegalArgumentException | NoSuchMethodException | SecurityException e) {
+      throw new RuntimeException(e);
+    }
+    BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.put(fnClass, constructor);
+    return constructor;
+  }
+
+  /** Generates a {@link DoFnInvoker} class for the given {@link DoFnSignature}. */
+  private static Class<? extends DoFnInvoker<?, ?>> generateInvokerClass(DoFnSignature signature) {
+    Class<? extends DoFn> fnClass = signature.getFnClass();
+
+    final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(fnClass);
+
+    DynamicType.Builder<?> builder =
+        new ByteBuddy()
+            // Create subclasses inside the target class, to have access to
+            // private and package-private bits
+            .with(
+                new NamingStrategy.SuffixingRandom("auxiliary") {
+                  @Override
+                  public String subclass(TypeDescription.Generic superClass) {
+                    return super.name(clazzDescription);
+                  }
+                })
+            // Create a subclass of DoFnInvoker
+            .subclass(DoFnInvoker.class, ConstructorStrategy.Default.NO_CONSTRUCTORS)
+            .defineField(
+                FN_DELEGATE_FIELD_NAME, fnClass, Visibility.PRIVATE, FieldManifestation.FINAL)
+            .defineConstructor(Visibility.PUBLIC)
+            .withParameter(fnClass)
+            .intercept(new InvokerConstructor())
+            // Delegate processElement(), startBundle() and finishBundle() to the fn.
+            .method(ElementMatchers.named("invokeProcessElement"))
+            .intercept(new ProcessElementDelegation(signature.getProcessElement()))
+            .method(ElementMatchers.named("invokeStartBundle"))
+            .intercept(
+                signature.getStartBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getStartBundle()))
+            .method(ElementMatchers.named("invokeFinishBundle"))
+            .intercept(
+                signature.getFinishBundle() == null
+                    ? new NoopMethodImplementation()
+                    : new BundleMethodDelegation(signature.getFinishBundle()))
+            .method(ElementMatchers.named("invokeSetup"))
+            .intercept(
+                signature.getSetup() == null
+                    ? new NoopMethodImplementation()
+                    : new LifecycleMethodDelegation(signature.getSetup()))
+            .method(ElementMatchers.named("invokeTeardown"))
+            .intercept(
+                signature.getTeardown() == null
+                    ? new NoopMethodImplementation()
```

Comment:
I'm refactoring this to something slightly nicer in the SplittableDoFn PR anyway.